### PR TITLE
Fixes breaking changes from Google on trends api using curl_cffi

### PR DESCRIPTION
**Pull request by maxwnewcomer**
Working pr for issue #561.

TODO: 
Update requirements.txt, better documentation, retry-abilities, normal tests, and proxy testing.

### CODE REVIEW

**Comment by maxwnewcomer** [comment]
Utilizes cURL impersonate as suggested by @nicktba to get around 429 errors. Still WIP. To use, clone my branch, do normal environment stuff, then `pip3 install curl_cffi`.

File: pytrends/request.py [hunk 1]
```diff
@@ -9,11 +10,29 @@
...
 from pytrends import exceptions
 
+from enum import Enum
 from urllib.parse import quote
 
 
 BASE_TRENDS_URL = 'https://trends.google.com/trends'
 
+# Taken from curl_cffi broswer type enum (https://github.com/yifeikong/curl_cffi/blob/master/curl_cffi/requests/session.py)
+class BrowserType(str, Enum):
```
Comment (williamegomezo) [inline pytrends/request.py:20]: But you dont have to copy and paste, just import if from: curl_cffi

from curl_cffi.requests.session import BrowserType

Comment (maxwnewcomer) [reply pytrends/request.py:20]: Lol very true, just was quick prototyping so people could use fix asap. Couldn't convert pr to draft like I intended. 

I'll fix in next iteration. thanks

**Comment by maxwnewcomer** [comment]
@williamegomezo would you like me to check this working with proxies before final review? Anything else you'd like to see?

**Comment by jesvinc** [comment]
I think this might be overkill. See my comment on the issue [here](https://github.com/GeneralMills/pytrends/issues/561#issuecomment-1462899426). If you change the request in `GetGoogleCookie` from a `GET` to a `POST`, the issue goes away. I implemented and tested this locally.

**Comment by maxwnewcomer** [comment]
> I think this might be overkill. See my comment on the issue [here](https://github.com/GeneralMills/pytrends/issues/561#issuecomment-1462899426). If you change the request in `GetGoogleCookie` from a `GET` to a `POST`, the issue goes away. I implemented and tested this locally.

You want me to add into this pr or do you want to add a separate one?

As mentioned in issue comments I think adding the curl impersonate might help future proof for future api updates and adds a new feature.

File: pytrends/request.py [hunk 2]
```diff
@@ -72,11 +80,19 @@ def GetGoogleCookie(self):
...
-                    ).cookies.items()))
+                    if self.impersonate:
+                        return dict(filter(lambda i: i[0] == 'NID', curl_requests.get(
+                            f'{BASE_TRENDS_URL}/?geo={self.hl[-2:]}',
+                            timeout=self.timeout,
+                            impersonate=self.impersonate,
+                            **self.requests_args
+                        ).cookies.items()))
+                    else:
+                        return dict(filter(lambda i: i[0] == 'NID', requests.get(
```
Comment (jesvinc) [inline pytrends/request.py:91]: Change from `requests.get` to `requests.post`.

File: pytrends/request.py [hunk 3]
```diff
@@ -85,12 +101,21 @@ def GetGoogleCookie(self):
...
                     proxy = ''
                 try:
-                    return dict(filter(lambda i: i[0] == 'NID', requests.get(
-                        f'{BASE_TRENDS_URL}/?geo={self.hl[-2:]}',
-                        timeout=self.timeout,
-                        proxies=proxy,
-                        **self.requests_args
-                    ).cookies.items()))
+                    if self.impersonate:
+                        return dict(filter(lambda i: i[0] == 'NID', curl_requests.get(
```
Comment (jesvinc) [inline pytrends/request.py:105]: Change from requests.get to requests.post.

Comment (maxwnewcomer) [reply pytrends/request.py:105]: The only weird thing is that the curl_requests.get() works fine in my tests, but the requests.get() doesn't.

File: pytrends/request.py [hunk 4]
```diff
@@ -85,12 +101,21 @@ def GetGoogleCookie(self):
...
+                            f'{BASE_TRENDS_URL}/?geo={self.hl[-2:]}',
+                            timeout=self.timeout,
+                            proxies=proxy,
+                            impersonate=self.impersonate,
+                            **self.requests_args
+                        ).cookies.items()))
+                    else:
+                        return dict(filter(lambda i: i[0] == 'NID', requests.get(
```
Comment (jchandy-xm) [inline pytrends/request.py:113]: Change from requests.get to requests.post.

File: pytrends/request.py (hunk 3 above)
Comment (jchandy-xm) [reply pytrends/request.py:105]: Oops yeah I commented on the wrong line. I agree, keep the `get` for curl.

Comment (jesvinc) [reply pytrends/request.py:105]: Oh oops, commented on the wrong line. Yeah ignore this one.

File: pytrends/request.py (hunk 4 above)
Comment (jesvinc) [inline pytrends/request.py:113]: Change from `requests.get` to `requests.post`.

**Comment by jesvinc** [comment]
@maxwnewcomer Did the tests work before? I pulled the repo and ran `pytest` and the tests still didn't pass.

**Comment by maxwnewcomer** [comment]
> @maxwnewcomer Did the tests work before? I pulled the repo and ran `pytest` and the tests still didn't pass.

When I do the requests.get -> .post the tests go from successful to failing.

Which sounds odd, because you would expect that if the tests are passing then we wouldn't be getting 429 errors in normal use.

**Comment by ckosmic** [comment]
Change the URL from `https://trends.google.com/trends/?geo=` to `https://trends.google.com/trends/explore?geo=` and it'll work with a get request

**Comment by jchandy-xm** [comment]
> Change the URL from `https://trends.google.com/trends/?geo=` to `https://trends.google.com/trends/explore/?geo=` and it'll work with a get request

Nice find, that works for me too.

**Comment by maxwnewcomer** [comment]
> Change the URL from `https://trends.google.com/trends/?geo=` to `https://trends.google.com/trends/explore/?geo=` and it'll work with a get request

The explore change does help more tests pass. Really should add some documentation about really making sure `--record-mode=rewrite` is enabled. Confused me a lot when running these. Seems like having the cookie pre-baked into tests is a bad idea (considering getting the cookie is actually the one of the problems with googles recent changes)

**Comment by gdavoian** [comment]
@maxwnewcomer at least for me, changing the request method from GET to POST completely fixed the issue, although I'm not sure why. I'm currently using the following **temporary workaround** in my own code to make everything work as before (attaching it here just as an example):

```python
import contextlib

import requests
from pytrends.request import TrendReq


class _TrendReq(TrendReq):
    ....

    def GetGoogleCookie(self):
        # TODO: make sure to get rid of this dirty hack
        with _requests_get_as_post():
            return super().GetGoogleCookie()


@contextlib.contextmanager
def _requests_get_as_post():
    requests.get, requests_get = requests.post, requests.get
    try:
        yield
    finally:
        requests.get = requests_get
```

The impersonate functionality might be useful in the future, but currently, I don't think it's urgently needed.

**Comment by maxwnewcomer** [comment]
> @maxwnewcomer at least for me, changing the request method from GET to POST completely fixed the issue, although I'm not sure why. I'm currently using the following **temporary workaround** in my own code to make everything work as before (attaching it here just as an example):
> 
> ```python
> import contextlib
> 
> import requests
> from pytrends.request import TrendReq
> 
> 
> class _TrendReq(TrendReq):
>     ....
> 
>     def GetGoogleCookie(self):
>         # TODO: make sure to get rid of this dirty hack
>         with _requests_get_as_post():
>             return super().GetGoogleCookie()
> 
> 
> @contextlib.contextmanager
> def _requests_get_as_post():
>     requests.get, requests_get = requests.post, requests.get
>     try:
>         yield
>     finally:
>         requests.get = requests_get
> ```
> 
> The impersonate functionality might be useful in the future, but currently, I don't think it's urgently needed.

The initial change suggested by @jesvinc didn't seem to work for me, but I'll try your implementation and see what happens. The impersonate seems to consistantly work for me, even though it's a still using `get` instead of `post` which just feels wrong.

**Comment by totencrab** [comment]
> Change the URL from `https://trends.google.com/trends/?geo=` to `https://trends.google.com/trends/explore/?geo=` and it'll work with a get request

This change seems to work for me.

**Comment by HectorSVAI** [comment]
After the cookies fix it works for a while with proxies until I get continuously 429 responses. With curl-impersonation, I get Failed to perform, ErrCode: 56, Reason: 'Received HTTP code 407 from proxy after CONNECT'

**Comment by HectorSVAI** [comment]
With proxies, either with cookies fix or curl impersonation it works a bit and then I get only 429 responses. USER_TYPE_SCRAPPER in response with impersonate="chrome110" for example.

**Comment by wglade** [comment]
Just wanted to say I made the change to the URL in the request.py file and that worked for me. I am running it without proxies because I haven't been able to get those working yet. Will definitely let you all know if I can get the curl impersonation working.

**Comment by jamesbwilson** [comment]
Hi, I'm curious if this fix is stalled? A generic pytrends install still will return 429s without modifying the library with one of the many suggested fixes.

**Comment by emlazzarin** [comment]
I believe this is resolved now but please let me know if that isn't true.

**Comment by josh-ashkinaze** [comment]
Not resolved

**Comment by Raidus** [comment]
I am experiencing even in the browser 429 issues. Maybe it's just a temporary issue and not related to pytrends

**Pull request closed without merging**